### ISSUE TITLE
Error for ImageLayer creation for an image with channels, scale, transform

### ISSUE THREAD

**Issue by thewtex** [napari/napari#2973]
## 🐛 Bug

After upgrading napari, a traceback is now obtained for an ImageLayer created with a channel dimension, `scale`, and `translate`.

## To Reproduce

Test script:

```
import napari
import numpy as np

data = np.random.rand(20, 20, 2).astype(np.float32)
image_layer = napari.layers.Image(data, scale=[1.0, 1.0], translate=[3.0, 4.0])
```

Results in the traceback:

```
Traceback (most recent call last):
  File "/tmp/napari_test.py", line 5, in <module>
    image_layer = napari.layers.Image(data, scale=[1.0, 1.0], translate=[3.0, 4.0])
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/image/image.py", line 290, in __init__
    self._update_dims()
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/base/base.py", line 544, in _update_dims
    self.refresh()
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/base/base.py", line 946, in refresh
    self.set_view_slice()
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/base/base.py", line 810, in set_view_slice
    self._set_view_slice()
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/image/image.py", line 519, in _set_view_slice
    indices = np.array(self._slice_indices)
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/base/base.py", line 612, in _slice_indices
    inv_transform = self._data_to_world.inverse
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/layers/base/base.py", line 996, in _data_to_world
    return self._transforms[1:3].simplified
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/utils/transforms/transforms.py", line 125, in simplified
    return tz.pipe(self[0], *[tf.compose for tf in self[1:]])
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/toolz/functoolz.py", line 627, in pipe
    data = func(data)
  File "/home/matt/bin/miniconda/envs/napari/lib/python3.9/site-packages/napari/utils/transforms/transforms.py", line 460, in compose
    affine_matrix = self.affine_matrix @ transform.affine_matrix
ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 4)
```

## Expected behavior

No traceback. Identified in CI: https://github.com/InsightSoftwareConsortium/napari-itk-io/pull/3

## Environment

 - Please copy and paste the information at napari info option in help menubar here:

```
napari: 0.4.10
Platform: Linux-5.9.8-050908-generic-x86_64-with-glibc2.31
System: Ubuntu 20.04.2 LTS
Python: 3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 05:02:46) [GCC 9.3.0]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.2
SciPy: 1.6.2
Dask: 2021.04.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6.0 NVIDIA 460.80
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 3840x2160, scale 1.0
- screen 2: resolution 3840x2160, scale 1.0

Plugins:
- console: 0.0.3
- napari-itk-io: 0.1.1.dev4+g2874092
- svg: 0.1.4
```

**Comment by jni** [comment]
Thanks @thewtex! I expect that this is related to #2967 and #2855, but certainly more dire! :grimacing:  We'll investigate and add a similar test of our own. :grimacing:

**Comment by sofroniewn** [comment]
I think here because the data is `(20, 20, 2)` we think it is 3D - not quite sure what to do though, if the data had been `(2, 20, 20)` would be have broadcast right? Maybe @andy-sweet can look into this too, especially if there was a regression

**Comment by andy-sweet** [comment]
Happy to take a look, but probably tomorrow.

**Comment by andy-sweet** [comment]
I'm pretty certain this was broken by [PR #2855](https://github.com/napari/napari/pull/2855) because it introduced two separate transforms (data2physical and physical2world) that contribute to `Layer._data_to_world`. These are not guaranteed to have matching dimensions because `Affine.__init__` infers the dimensionality of the transform as the highest dimensionality of the transform component inputs, whereas the other transforms in `TransformChain` inherit the dimensionality of the layer. I suspect that even before that PR, the same example above would generated a `TransformChain` with mixed dimensionality, but we never called `TransformChain.simplified` to cause the bug.

Here, the physical2world transform is inferred to be 2D, but the layer is inferred to be 3D because in general we don't distinguish spatial dimensions from channels.

I think the simplest and best fix is to add an optional keyword `ndim` argument to `Affine.__init__` because there's no need to infer the dimensionality when we know what is from `Layer`. This bug can be fixed in [PR #2990](https://github.com/napari/napari/pull/2990) because that adds `CompositeAffine` which has an `ndim` argument as described, so if we can wait another day or so, it might be worth just waiting for that to be merged. Otherwise, I can make a hot-fix PR.

**Comment by sofroniewn** [comment]
> This bug can be fixed in PR #2990 because that adds CompositeAffine which has an ndim argument as described, so if we can wait another day or so, it might be worth just waiting for that to be merged. Otherwise, I can make a hot-fix PR.

I think ok to wait, that PR will go in before next release, so i think its ok

**Comment by andy-sweet** [comment]
Note that even with the fix in [PR #2990](https://github.com/napari/napari/pull/2990), the 2D scale and translate vectors will be padded with a leading 1 and 0 respectively to become `(1, 1, 1)` and `(0, 3, 4)`, so the resulting translate vector probably isn't what we want where the non-spatial channel dimension is last.

**Comment by jni** [comment]
Yeah I hadn't read the above correctly — indeed @thewtex for that image you will *have* to provide `channel_axis=-1` as a kwarg to `add_image`. There is otherwise no way for us to know that the last axis is channels. (yes yes, ITK and typing is so good, we know. :joy:)

**Comment by andy-sweet** [comment]
The crash should be fixed by PR #2990, so closing this issue.

**Comment by thewtex** [comment]
@andy-sweet @jni @sofroniewn thank you! :100:
